### PR TITLE
SDHCI refactor

### PR DESCRIPTION
**Pull request by Deryugin**
* Refactor SDHCI driver
* Add driver for i.MX6 SD driver
* Add clocks for i.MX6 sdhc

### CODE REVIEW

File: src/drivers/mmc/host/sdhci/sdhci.c [hunk 1]
```diff
@@ -6,301 +6,30 @@
...
-		mmc->priv = host;
-		mmc_sw_reset(mmc);
-	}
-#endif /* !NEW_API */
-
-	_reg_dump(host);
-
-	return 0;
-}
+PERIPH_MEMORY_DEFINE(imx6_usdhc, BASE_ADDR, 0x4000 * MAX_HC_NUM);
```
Comment (anton-bondarev) [inline src/drivers/mmc/host/sdhci/sdhci.c]: May be use macro
`#define IO_REG_SIZE 0x4000`

File: src/drivers/mmc/host/sdhci/sdhci_ops.c [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+	int timeout = TIMEOUT;
+	while(timeout--);
+
+	if (sdhci_readl(host, USDHC_SYS_CTRL) & USDHC_SYS_CTRL_RSTA) {
+		log_error("Reset timeout");
+	}
+
+	sdhci_writel(host, USDHC_VEND_SPEC,
+			sdhci_readl(host, USDHC_VEND_SPEC) | (1 << 31) | (1 << 14) | (1 << 13));
```
Comment (anton-bondarev) [inline src/drivers/mmc/host/sdhci/sdhci_ops.c]: Set macro for bits?

File: src/drivers/mmc/host/sdhci/sdhci_ops.c [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	tmp = 512 | (1 << 16);
+	sdhci_writel(host, USDHC_BLK_ATT, tmp);
+	if (sdhci_readl(host, USDHC_PRES_STATE) & USDHC_PRES_STATE_CDPL) {
+		mmc = mmc_alloc_host();
+		mmc->ops = &sdhci_mmc_ops;
+		mmc->priv = host;
+		mmc_sw_reset(mmc);
+		sdhci_reg_dump(host);
+	} else {
+		log_debug("CDPL bit is zero! %08x", sdhci_readl(host, USDHC_PRES_STATE));
```
Comment (alexkalmuk) [inline src/drivers/mmc/host/sdhci/sdhci_ops.c]: Is this situation properly handled? The upper layer will get 0 from this function, but mmc will be not allocated, it's OK?

File: src/drivers/mmc/host/sdhci/sdhci_ops.c [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+static void sdhci_reset(struct sdhci_host *host) {
+	sdhci_writel(host, USDHC_SYS_CTRL, sdhci_readl(host, USDHC_SYS_CTRL) | USDHC_SYS_CTRL_RSTA);
+
+	int timeout = TIMEOUT;
+	while(timeout--);
+
+	if (sdhci_readl(host, USDHC_SYS_CTRL) & USDHC_SYS_CTRL_RSTA) {
+		log_error("Reset timeout");
```
Comment (alexkalmuk) [inline src/drivers/mmc/host/sdhci/sdhci_ops.c]: In case if reset is unsuccessful how is it possible to continue?

Comment (alexkalmuk) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: You use this function in probe, which seems to be broken if this one failed

File: src/drivers/mmc/host/sdhci/sdhci_ops.c [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+int sdhci_hw_probe(struct sdhci_host *host) {
+	uint32_t tmp;
+	struct mmc_host *mmc;
+
+	sdhci_reset(host);
+
+	sdhci_writel(host, USDHC_SYS_CTRL, 0xb011f);
+	sdhci_writel(host, USDHC_INT_STATUS_EN, -1);
+	sdhci_writel(host, USDHC_PROT_CTRL, sdhci_readl(host, USDHC_PROT_CTRL) | 1 | 4);
```
Comment (alexkalmuk) [inline src/drivers/mmc/host/sdhci/sdhci_ops.c]: Why do not use hex?

File: src/drivers/mmc/host/sdhci/sdhci_ops.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	sdhci_writel(host, USDHC_CMD_XFR_TYP, wcmd);
+
+	/* Wait command to be completed */
+	while(!(sdhci_readl(host, USDHC_INT_STATUS) & 0x1) && retry > 0) {
+		usleep(USEC_PER_MSEC);
+		retry--;
+	}
+
+	if (retry == 0) {
+		log_debug("Timeout!");
```
Comment (alexkalmuk) [inline src/drivers/mmc/host/sdhci/sdhci_ops.c]: log_error and return? Or timeout is normal situation?

Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: Fixed

File: src/drivers/mmc/host/sdhci/sdhci_ops.c (hunk 4 above)
Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: Fixed

File: src/drivers/mmc/host/sdhci/sdhci_ops.c (hunk 5 above)
Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: Fixed

File: src/drivers/mmc/host/sdhci/sdhci_ops.c (hunk 3 above)
Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: In this case there's no card inserted, so yes, it's ok not to alloc MMC

File: src/drivers/mmc/host/sdhci/sdhci_ops.c (hunk 2 above)
Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci_ops.c]: Those bits are vendor-specific, so they can't be described for all the controllers. As far as I remember, originally this value was taken from u-boot for some specific i.MX6 board

File: src/drivers/mmc/host/sdhci/sdhci.c (hunk 1 above)
Comment (Deryugin) [reply src/drivers/mmc/host/sdhci/sdhci.c]: Done

**Pull request merged**